File: examples/ejemplo 6 - kafka-serialization/KafkaSerialization.java
import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.kafka.clients.producer.*;
import org.apache.kafka.common.serialization.Serializer;
import org.apache.kafka.common.serialization.StringSerializer;

import java.util.Properties;

/**
 * Clase de ejemplo para serialización
 */
class User {
    private String id;
    private String name;
    private int age;
    
    public User() {}
    
    public User(String id, String name, int age) {
        this.id = id;
        this.name = name;
        this.age = age;
    }
    
    // Getters y setters
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    public String getName() { return name; }
    public void setName(String name) { this.name = name; }
    public int getAge() { return age; }
    public void setAge(int age) { this.age = age; }
}

/**
 * Serializador personalizado para User usando JSON
 */
class UserSerializer implements Serializer<User> {
    private final ObjectMapper objectMapper = new ObjectMapper();
    
    @Override
    public byte[] serialize(String topic, User data) {
        try {
            return objectMapper.writeValueAsBytes(data);
        } catch (Exception e) {
            throw new RuntimeException("Error al serializar User", e);
        }
    }
}

/**
 * Ejemplo de serialización personalizada en Kafka
 */
public class KafkaSerialization {
    
    public static void main(String[] args) {
        Properties props = new Properties();
        props.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092");
        props.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class.getName());
        props.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, UserSerializer.class.getName());
        props.put(ProducerConfig.ACKS_CONFIG, "all");
        
        KafkaProducer<String, User> producer = new KafkaProducer<>(props);
        
        try {
            String topic = "users-topic";
            
            // Crear y enviar usuarios
            User[] users = {
                new User("1", "Alice", 30),
                new User("2", "Bob", 25),
                new User("3", "Charlie", 35)
            };
            
            for (User user : users) {
                ProducerRecord<String, User> record = 
                    new ProducerRecord<>(topic, user.getId(), user);
                
                producer.send(record, (metadata, exception) -> {
                    if (exception == null) {
                        System.out.printf("Usuario enviado: %s (Partición: %d, Offset: %d)%n",
                            user.getName(), metadata.partition(), metadata.offset());
                    } else {
                        System.err.println("Error: " + exception.getMessage());
                    }
                });
            }
            
            System.out.println("\n✓ Todos los usuarios serializados y enviados!");
            System.out.println("Los objetos User fueron convertidos a JSON automáticamente.");
            
        } finally {
            producer.flush();
            producer.close();
            System.out.println("Productor cerrado.");
        }
    }
}

File: examples/ejemplo 6 - kafka-serialization/KafkaSerialization.kt
import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.kafka.clients.producer.KafkaProducer
import org.apache.kafka.clients.producer.ProducerConfig
import org.apache.kafka.clients.producer.ProducerRecord
import org.apache.kafka.common.serialization.Serializer
import org.apache.kafka.common.serialization.StringSerializer
import java.util.Properties

/**
 * Clase de ejemplo para serialización
 */
data class User(
    val id: String,
    val name: String,
    val age: Int
)

/**
 * Serializador personalizado para User usando JSON
 */
class UserSerializer : Serializer<User> {
    private val objectMapper = ObjectMapper()
    
    override fun serialize(topic: String?, data: User?): ByteArray? {
        return try {
            data?.let { objectMapper.writeValueAsBytes(it) }
        } catch (e: Exception) {
            throw RuntimeException("Error al serializar User", e)
        }
    }
}

/**
 * Ejemplo de serialización personalizada en Kafka con Kotlin
 */
fun main() {
    val props = Properties().apply {
        put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, "localhost:9092")
        put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer::class.java.name)
        put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, UserSerializer::class.java.name)
        put(ProducerConfig.ACKS_CONFIG, "all")
    }
    
    KafkaProducer<String, User>(props).use { producer ->
        try {
            val topic = "users-topic"
            
            // Crear y enviar usuarios
            val users = listOf(
                User("1", "Alice", 30),
                User("2", "Bob", 25),
                User("3", "Charlie", 35)
            )
            
            users.forEach { user ->
                val record = ProducerRecord(topic, user.id, user)
                
                producer.send(record) { metadata, exception ->
                    if (exception == null) {
                        println("Usuario enviado: ${user.name} (Partición: ${metadata.partition()}, Offset: ${metadata.offset()})")
                    } else {
                        System.err.println("Error: ${exception.message}")
                    }
                }
            }
            
            println("\n✓ Todos los usuarios serializados y enviados!")
            println("Los objetos User fueron convertidos a JSON automáticamente.")
            
        } finally {
            producer.flush()
            println("Productor cerrado.")
        }
    }
}
